Added support for Alarm notifications & positions history cleaner & show alarms in report & Mute button & unlock UI files (for windows)

ninioe opened the pull request:
1. added support for Alarm and Alarm Type in popup and also in history
records. can be muted with a toggle button in the UI
2. added a timer to clear positions history once a day. the default is
to save positions history for 7 days if not defined in the configuration
file.
3. prevent the lock of the UI files (js, html, css, etc..) for wondows
developers in debug mode. It's easier to do changes to the UI without
stopping the app each time.
4. tools: added support in test-generator.py to simulate Alert, also
added minify.bat file to compile new js files for the UI using sencha
SDK for windows developers

tananaev commented on debug.xml:
```diff
@@ -65,10 +65,15 @@
-    <entry key='database.password'></entry>
+    <entry key='database.password'></entry>-->
+
+    <entry key='database.driver'>com.mysql.jdbc.Driver</entry>
```
Please exclude this change.

ninioe replied:
how do I exclude a change in a file?

On Wed, Jul 13, 2016 at 1:46 AM, Anton Tananaev <notifications@github.com>
wrote:

> In debug.xml
> <https://github.com/tananaev/traccar/pull/2104#discussion_r70536399>:
>
> >      <entry key='database.url'>jdbc:h2:./target/database</entry>
> >      <entry key='database.user'>sa</entry>
> > -    <entry key='database.password'></entry>
> > +    <entry key='database.password'></entry>-->
> > +
> > +    <entry key='database.driver'>com.mysql.jdbc.Driver</entry>
>
> Please exclude this change.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/tananaev/traccar/pull/2104/files/d045c1e2174146ddc8ad8b2ef4442863f3d84a11#r70536399>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AH9fF1YfAWofjwO44_zCIi_9dqsseieJks5qVBkrgaJpZM4JKXy->
> .
>

tananaev replied:
Just undo it.

ninioe replied:
I prefer to use mysql instead of h2.
do I have to delete this entry from my debug.xml file before I commit and
them put it back?

On Wed, Jul 13, 2016 at 7:55 AM, Anton Tananaev <notifications@github.com>
wrote:

> In debug.xml
> <https://github.com/tananaev/traccar/pull/2104#discussion_r70564269>:
>
> >      <entry key='database.url'>jdbc:h2:./target/database</entry>
> >      <entry key='database.user'>sa</entry>
> > -    <entry key='database.password'></entry>
> > +    <entry key='database.password'></entry>-->
> > +
> > +    <entry key='database.driver'>com.mysql.jdbc.Driver</entry>
>
> Just undo it.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/tananaev/traccar/pull/2104/files/d045c1e2174146ddc8ad8b2ef4442863f3d84a11#r70564269>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AH9fF_k1kZ-DgIAHY3zPSTlWMQiJnSYrks5qVG-3gaJpZM4JKXy->
> .
>

tananaev replied:
Yes, you need to remove it first.

ninioe replied:
Ok I will do all the changes and send new pull request

‫ב-13 ביולי 2016, בשעה 8:22, ‏‏Anton Tananaev ‏<notifications@github.com> כתב/ה:‬

> In debug.xml:
> 
> >      <entry key='database.url'>jdbc:h2:./target/database</entry>
> >      <entry key='database.user'>sa</entry>
> > -    <entry key='database.password'></entry>
> > +    <entry key='database.password'></entry>-->
> > +
> > +    <entry key='database.driver'>com.mysql.jdbc.Driver</entry>
> Yes, you need to remove it first.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

tananaev commented on src/org/traccar/Main.java:
```diff
@@ -35,6 +36,20 @@ public static void main(String[] args) throws Exception {
             Context.getWebServer().start();
         }
 
+        //added by Erez
```
Remove all comments like this.

tananaev commented on src/org/traccar/Main.java:
```diff
@@ -35,6 +36,20 @@ public static void main(String[] args) throws Exception {
+                    Log.warning(error);
+                }
+            }
+        }, 10*1000, 24*60*60*1000);
```
Move constants to static variables at the top of the class and give them some name.

tananaev commented on src/org/traccar/database/DataManager.java:
```diff
@@ -521,6 +521,31 @@ public void updateLatestPosition(Position position) throws SQLException {
+    //added by Erez
+    public void clearPositionsHistory() throws SQLException {
+        //SimpleDateFormat s = new SimpleDateFormat("yyyy-MM-dd");
+        String histDays = config.getString("database.positionsHistoryDays");
```
Use `getInteger` instead.

ninioe replied:
how do I use the getInteger, can you give me an example?

Integer.getInteger(histDays) cause an exeption


On Wed, Jul 13, 2016 at 1:48 AM, Anton Tananaev <notifications@github.com>
wrote:

> In src/org/traccar/database/DataManager.java
> <https://github.com/tananaev/traccar/pull/2104#discussion_r70536620>:
>
> > @@ -521,6 +521,31 @@ public void updateLatestPosition(Position position) throws SQLException {
> >                  .executeQuery(Position.class);
> >      }
> >
> > +    //added by Erez
> > +    public void clearPositionsHistory() throws SQLException {
> > +        //SimpleDateFormat s = new SimpleDateFormat("yyyy-MM-dd");
> > +        String histDays = config.getString("database.positionsHistoryDays");
>
> Use getInteger instead.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/tananaev/traccar/pull/2104/files/d045c1e2174146ddc8ad8b2ef4442863f3d84a11#r70536620>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AH9fFxlXC48_2YS79mEKzoeYgT6TJ01cks5qVBmjgaJpZM4JKXy->
> .
>

ninioe replied:
ok, I found how to do this in the code

On Wed, Jul 13, 2016 at 9:39 AM, Ninio Erez <ninioe@gmail.com> wrote:

> how do I use the getInteger, can you give me an example?
>
> Integer.getInteger(histDays) cause an exeption
>
>
> On Wed, Jul 13, 2016 at 1:48 AM, Anton Tananaev <notifications@github.com>
> wrote:
>
>> In src/org/traccar/database/DataManager.java
>> <https://github.com/tananaev/traccar/pull/2104#discussion_r70536620>:
>>
>> > @@ -521,6 +521,31 @@ public void updateLatestPosition(Position position) throws SQLException {
>> >                  .executeQuery(Position.class);
>> >      }
>> >
>> > +    //added by Erez
>> > +    public void clearPositionsHistory() throws SQLException {
>> > +        //SimpleDateFormat s = new SimpleDateFormat("yyyy-MM-dd");
>> > +        String histDays = config.getString("database.positionsHistoryDays");
>>
>> Use getInteger instead.
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/tananaev/traccar/pull/2104/files/d045c1e2174146ddc8ad8b2ef4442863f3d84a11#r70536620>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe/AH9fFxlXC48_2YS79mEKzoeYgT6TJ01cks5qVBmjgaJpZM4JKXy->
>> .
>>
>
>

tananaev commented on src/org/traccar/database/DataManager.java:
```diff
@@ -521,6 +521,31 @@ public void updateLatestPosition(Position position) throws SQLException {
+        if (histDays == null) {
+            histDays = "7";
+        }
+        int n = Integer.parseInt(histDays);
```
What is `n`? Please give a meaningful name.

tananaev commented on src/org/traccar/database/DataManager.java:
```diff
@@ -521,6 +521,31 @@ public void updateLatestPosition(Position position) throws SQLException {
+
+                Date dateBefore = new Date(lastUpdate.getTime() - n * 24 * 3600 * 1000 ); //Subtract n days
+                //String dt = s.format(dateBefore);
+                String sql = "DELETE FROM positions WHERE deviceid=:deviceId and servertime<:serverTime";
```
Query should be in the config. Also, how do you handle case when we remove device position that is referenced via `positionId` in devices?

tananaev commented on src/org/traccar/model/Position.java:
```diff
@@ -28,6 +28,7 @@
     public static final String KEY_GPS = "gps";
     public static final String KEY_EVENT = "event";
     public static final String KEY_ALARM = "alarm";
+    public static final String KEY_ALARM_TYPE = "alarm-type";   //added by Erez
```
No need for new key. Just use "alarm".

tananaev commented on src/org/traccar/protocol/H02ProtocolDecoder.java:
```diff
@@ -64,8 +64,21 @@ private static double readCoordinate(ChannelBuffer buf, boolean lon) {
             position.set(Position.KEY_ALARM, true);
+
+            if (!BitUtil.check(status, 0)){//added by Erez
+                position.set(Position.KEY_ALARM_TYPE, "theft alarm");
```
Word "alarm" can be removed because we already know that it's an alarm.

tananaev commented on src/org/traccar/web/WebServer.java:
```diff
@@ -118,6 +118,13 @@ private void initWebApp() {
         if (config.getBoolean("web.debug")) {
             resourceHandler.setWelcomeFiles(new String[] {"debug.html"});
+            //Troubleshooting Locked Files on Windows (changed by Erez)
+            resourceHandler.setMinMemoryMappedContentLength(-1);
```
What does this do? Also, please remove commented out code if it's not needed.

tananaev commented on tools/minify.bat:
```diff
@@ -0,0 +1,5 @@
+@echo off
+cd C:\Users\Erez\Documents\traccar\web
```
Please make it more generic.

tananaev commented on tools/test-generator.py:
```diff
@@ -32,8 +32,11 @@
+def send(lat, lon, course, alarm):
     params = (('id', id), ('timestamp', int(time.time())), ('lat', lat), ('lon', lon), ('bearing', course))
+    if alarm:
+        params = params + (('alarm', 'true'),)
```
Please remove comma if not required there.

tananaev commented:
There are quite a few comments.

Also, I think alarm should be displayed as a standard notification.

Please exclude web UI changes that you have there because they seem to be very specific to your use case. Possibly keep the color of the alarm, but the function looks very weird.

Abyss777 commented:
My two cents:
Better way to catch alarm is to do it via event handler. It is in my plans side by side with catching ignition on/off event.

We just need to make an agreement about position attributes names that will throw events.

Special thanks from me about 3 point, it is really annoying.

> the default is to save positions history for 7 days if not defined in the configuration file

The default should be forever I think.

ninioe commented:
Hi Anton,
I think you are right, the Alarms should be under the notifications, but
how does it work?
I have tried all kind of the notifications with "Send Via Web" but nothing
happens.

thanks
Erez

On Wed, Jul 13, 2016 at 2:10 AM, Anton Tananaev <notifications@github.com>
wrote:

> There are quite a few comments.
>
> Also, I think alarm should be displayed as a standard notification.
>
> Please exclude web UI changes that you have there because they seem to be
> very specific to your use case. Possibly keep the color of the alarm, but
> the function looks very weird.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/tananaev/traccar/pull/2104#issuecomment-232209179>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AH9fF9U41H9Z0XQfz9Wscc1ymZvA-Dzxks5qVB7cgaJpZM4JKXy->
> .
>

tananaev commented:
You can take a look at existing event handlers. For example, `OverspeedEventHandler`.

ninioe commented:
but how do I simulate an event?
I have set a geofense set all the notification via web checked and run
the test-generator.py but I don't get any notifications

On Wed, Jul 13, 2016 at 11:18 AM, Anton Tananaev <notifications@github.com>
wrote:

> You can take a look at existing event handlers. For example,
> OverspeedEventHandler.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/tananaev/traccar/pull/2104#issuecomment-232288868>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AH9fF-ZNx5tmzSmeM7MbIPKLnBw27k2jks5qVJ9qgaJpZM4JKXy->
> .
>

tananaev commented:
Have you linked geofence to your device?

ninioe commented:
Yes

‫ב-13 ביולי 2016, בשעה 13:13, ‏‏Anton Tananaev ‏<notifications@github.com> כתב/ה:‬

> Have you linked geofence to your device?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

ninioe commented:
It works now

Thanks

‫ב-13 ביולי 2016, בשעה 13:13, ‏‏Anton Tananaev ‏<notifications@github.com> כתב/ה:‬

> Have you linked geofence to your device?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

ninioe commented:
Hi,
1. Do you want me to remove the toggle button from the UI, or can we use it
to temporary mute all notifications?
2. Could we add the beep sound to the web notifications?
3. What about the history report, should I remove the Alarm Type? can you
add the notification inside the Alarm field?
4. How do I get all the latest changes you did with the GitHub app?

sorry for all those questions, I quit new in this but I'm learning and I
want to assist as much as I can in this project...

Thanks
Erez


On Wed, Jul 13, 2016 at 2:10 AM, Anton Tananaev <notifications@github.com>
wrote:

> There are quite a few comments.
>
> Also, I think alarm should be displayed as a standard notification.
>
> Please exclude web UI changes that you have there because they seem to be
> very specific to your use case. Possibly keep the color of the alarm, but
> the function looks very weird.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/tananaev/traccar/pull/2104#issuecomment-232209179>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AH9fF9U41H9Z0XQfz9Wscc1ymZvA-Dzxks5qVB7cgaJpZM4JKXy->
> .
>

tananaev commented:
1. Which toggle button? Please provide a screenshot.
2. If you want to keep it, it should definitely be configurable. And probably actual sound file should be separate from code.
3. Remove alarm from report for now. Better reporting system is coming at some point. We can add it back then. What do you mean by notification inside alarm field?
4. You need to fetch and merge the changes if you want to sync with upstream repo.

ninioe commented:
1. this toggle button:
[image: Inline image 1][image: Inline image 2]

2. there is no sound file, this function do the beep sound:

traccar\web\app\view\StateController.js

beep: function() {
    if(this.snd == null){
        this.snd = new
Audio("data:audio/wav;base64,//uQRAAAAWMSLwUIYAAsYkXgoQwAEaYLWfkWgAI0wWs/ItAAAGDgYtAgAyN+QWaAAihwMWm4G8QQRDiMcCBcH3Cc+CDv/7xA4Tvh9Rz/y8QADBwMWgQAZG/ILNAARQ4GLTcDeIIIhxGOBAuD7hOfBB3/94gcJ3w+o5/5eIAIAAAVwWgQAVQ2ORaIQwEMAJiDg95G4nQL7mQVWI6GwRcfsZAcsKkJvxgxEjzFUgfHoSQ9Qq7KNwqHwuB13MA4a1q/DmBrHgPcmjiGoh//EwC5nGPEmS4RcfkVKOhJf+WOgoxJclFz3kgn//dBA+ya1GhurNn8zb//9NNutNuhz31f////9vt///z+IdAEAAAK4LQIAKobHItEIYCGAExBwe8jcToF9zIKrEdDYIuP2MgOWFSE34wYiR5iqQPj0JIeoVdlG4VD4XA67mAcNa1fhzA1jwHuTRxDUQ//iYBczjHiTJcIuPyKlHQkv/LHQUYkuSi57yQT//uggfZNajQ3Vmz+Zt//+mm3Wm3Q576v////+32///5/EOgAAADVghQAAAAA//uQZAUAB1WI0PZugAAAAAoQwAAAEk3nRd2qAAAAACiDgAAAAAAABCqEEQRLCgwpBGMlJkIz8jKhGvj4k6jzRnqasNKIeoh5gI7BJaC1A1AoNBjJgbyApVS4IDlZgDU5WUAxEKDNmmALHzZp0Fkz1FMTmGFl1FMEyodIavcCAUHDWrKAIA4aa2oCgILEBupZgHvAhEBcZ6joQBxS76AgccrFlczBvKLC0QI2cBoCFvfTDAo7eoOQInqDPBtvrDEZBNYN5xwNwxQRfw8ZQ5wQVLvO8OYU+mHvFLlDh05Mdg7BT6YrRPpCBznMB2r//xKJjyyOh+cImr2/4doscwD6neZjuZR4AgAABYAAAABy1xcdQtxYBYYZdifkUDgzzXaXn98Z0oi9ILU5mBjFANmRwlVJ3/6jYDAmxaiDG3/6xjQQCCKkRb/6kg/wW+kSJ5//rLobkLSiKmqP/0ikJuDaSaSf/6JiLYLEYnW/+kXg1WRVJL/9EmQ1YZIsv/6Qzwy5qk7/+tEU0nkls3/zIUMPKNX/6yZLf+kFgAfgGyLFAUwY//uQZAUABcd5UiNPVXAAAApAAAAAE0VZQKw9ISAAACgAAAAAVQIygIElVrFkBS+Jhi+EAuu+lKAkYUEIsmEAEoMeDmCETMvfSHTGkF5RWH7kz/ESHWPAq/kcCRhqBtMdokPdM7vil7RG98A2sc7zO6ZvTdM7pmOUAZTnJW+NXxqmd41dqJ6mLTXxrPpnV8avaIf5SvL7pndPvPpndJR9Kuu8fePvuiuhorgWjp7Mf/PRjxcFCPDkW31srioCExivv9lcwKEaHsf/7ow2Fl1T/9RkXgEhYElAoCLFtMArxwivDJJ+bR1HTKJdlEoTELCIqgEwVGSQ+hIm0NbK8WXcTEI0UPoa2NbG4y2K00JEWbZavJXkYaqo9CRHS55FcZTjKEk3NKoCYUnSQ0rWxrZbFKbKIhOKPZe1cJKzZSaQrIyULHDZmV5K4xySsDRKWOruanGtjLJXFEmwaIbDLX0hIPBUQPVFVkQkDoUNfSoDgQGKPekoxeGzA4DUvnn4bxzcZrtJyipKfPNy5w+9lnXwgqsiyHNeSVpemw4bWb9psYeq//uQZBoABQt4yMVxYAIAAAkQoAAAHvYpL5m6AAgAACXDAAAAD59jblTirQe9upFsmZbpMudy7Lz1X1DYsxOOSWpfPqNX2WqktK0DMvuGwlbNj44TleLPQ+Gsfb+GOWOKJoIrWb3cIMeeON6lz2umTqMXV8Mj30yWPpjoSa9ujK8SyeJP5y5mOW1D6hvLepeveEAEDo0mgCRClOEgANv3B9a6fikgUSu/DmAMATrGx7nng5p5iimPNZsfQLYB2sDLIkzRKZOHGAaUyDcpFBSLG9MCQALgAIgQs2YunOszLSAyQYPVC2YdGGeHD2dTdJk1pAHGAWDjnkcLKFymS3RQZTInzySoBwMG0QueC3gMsCEYxUqlrcxK6k1LQQcsmyYeQPdC2YfuGPASCBkcVMQQqpVJshui1tkXQJQV0OXGAZMXSOEEBRirXbVRQW7ugq7IM7rPWSZyDlM3IuNEkxzCOJ0ny2ThNkyRai1b6ev//3dzNGzNb//4uAvHT5sURcZCFcuKLhOFs8mLAAEAt4UWAAIABAAAAAB4qbHo0tIjVkUU//uQZAwABfSFz3ZqQAAAAAngwAAAE1HjMp2qAAAAACZDgAAAD5UkTE1UgZEUExqYynN1qZvqIOREEFmBcJQkwdxiFtw0qEOkGYfRDifBui9MQg4QAHAqWtAWHoCxu1Yf4VfWLPIM2mHDFsbQEVGwyqQoQcwnfHeIkNt9YnkiaS1oizycqJrx4KOQjahZxWbcZgztj2c49nKmkId44S71j0c8eV9yDK6uPRzx5X18eDvjvQ6yKo9ZSS6l//8elePK/Lf//IInrOF/FvDoADYAGBMGb7FtErm5MXMlmPAJQVgWta7Zx2go+8xJ0UiCb8LHHdftWyLJE0QIAIsI+UbXu67dZMjmgDGCGl1H+vpF4NSDckSIkk7Vd+sxEhBQMRU8j/12UIRhzSaUdQ+rQU5kGeFxm+hb1oh6pWWmv3uvmReDl0UnvtapVaIzo1jZbf/pD6ElLqSX+rUmOQNpJFa/r+sa4e/pBlAABoAAAAA3CUgShLdGIxsY7AUABPRrgCABdDuQ5GC7DqPQCgbbJUAoRSUj+NIEig0YfyWUho1VBBBA//uQZB4ABZx5zfMakeAAAAmwAAAAF5F3P0w9GtAAACfAAAAAwLhMDmAYWMgVEG1U0FIGCBgXBXAtfMH10000EEEEEECUBYln03TTTdNBDZopopYvrTTdNa325mImNg3TTPV9q3pmY0xoO6bv3r00y+IDGid/9aaaZTGMuj9mpu9Mpio1dXrr5HERTZSmqU36A3CumzN/9Robv/Xx4v9ijkSRSNLQhAWumap82WRSBUqXStV/YcS+XVLnSS+WLDroqArFkMEsAS+eWmrUzrO0oEmE40RlMZ5+ODIkAyKAGUwZ3mVKmcamcJnMW26MRPgUw6j+LkhyHGVGYjSUUKNpuJUQoOIAyDvEyG8S5yfK6dhZc0Tx1KI/gviKL6qvvFs1+bWtaz58uUNnryq6kt5RzOCkPWlVqVX2a/EEBUdU1KrXLf40GoiiFXK///qpoiDXrOgqDR38JB0bw7SoL+ZB9o1RCkQjQ2CBYZKd/+VJxZRRZlqSkKiws0WFxUyCwsKiMy7hUVFhIaCrNQsKkTIsLivwKKigsj8XYlwt/WKi2N4d//uQRCSAAjURNIHpMZBGYiaQPSYyAAABLAAAAAAAACWAAAAApUF/Mg+0aohSIRobBAsMlO//Kk4soosy1JSFRYWaLC4qZBYWFRGZdwqKiwkNBVmoWFSJkWFxX4FFRQWR+LsS4W/rFRb/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////VEFHAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAU291bmRib3kuZGUAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAMjAwNGh0dHA6Ly93d3cuc291bmRib3kuZGUAAAAAAAAAACU=");
    }
    this.snd.play();
},

3. I meant the Event/Alarm

[image: Inline image 4]


Thanks

Erez






On Wed, Jul 13, 2016 at 4:08 PM, Anton Tananaev <notifications@github.com>
wrote:

>
>    1. Which toggle button? Please provide a screenshot.
>    2. If you want to keep it, it should definitely be configurable. And
>    probably actual sound file should be separate from code.
>    3. Remove alarm from report for now. Better reporting system is coming
>    at some point. We can add it back then. What do you mean by notification
>    inside alarm field?
>    4. You need to fetch and merge the changes if you want to sync with
>    upstream repo.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/tananaev/traccar/pull/2104#issuecomment-232349144>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AH9fF2opYTssJHGFrBp1WsGqFk5gxn6wks5qVONPgaJpZM4JKXy->
> .
>

tananaev commented:
I would recommend to answer through GitHub web interface. Your images are not visible and there is no way to properly format the code.

As for sound file, your "function" contains base64-encoded binary file with sound. That's exactly what I'm talking about. It shouldn't be there.

ninioe commented:
this is the toggle button I meant, it can be used to mute the notifications:
![4a16ba95-b9de-4919-b17e-49e9c7c9e734](https://cloud.githubusercontent.com/assets/8347415/16827910/8ed3d7f6-4993-11e6-8416-6e35ea92770e.png)
![22fac371-663c-43d2-84d2-de3ed4d09a6a](https://cloud.githubusercontent.com/assets/8347415/16827909/8ed0da38-4993-11e6-918f-805298e7191a.png)

and this is my suggestion for the report:
![unnamed](https://cloud.githubusercontent.com/assets/8347415/16827935/cc183418-4993-11e6-8d60-58d7f0e9de3b.png)

as for the beep sound do you prefer to use a sound file instead the base64?

ninioe commented:
can you add standard notification for SOS and vibration alarms?

tananaev commented:
Just add them yourself. It should be fairly easy.

As for alarm toggle button, I would recommend to save it in user settings, so user preference saved in the database.

ninioe commented:
How do I show a notification message?

As for the mute button, I don't think it is necessary to save the state to the data base

‫ב-14 ביולי 2016, בשעה 7:46, ‏‏Anton Tananaev ‏<notifications@github.com> כתב/ה:‬

> Just add them yourself. It should be fairly easy.
> 
> As for alarm toggle button, I would recommend to save it in user settings, so user preference saved in the database.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

tananaev commented:
You need to send notification from the back-end and it will show up automatically. Just need to add new type.

As for button, I guess we can add without saving, but then it should be OFF by default.

ninioe commented:
I can't use a notification if it's not exist in the notifications table in
the data base
can you add a general Alarms to the notifications with default selected via
web for the users?
or should I bypass the getUserNotificationByType function inside
updateEvent function
in NotificationManager class?

another way is to check the alarm attribute on the client side and call
Ext.toast

On Thu, Jul 14, 2016 at 9:24 AM, Anton Tananaev <notifications@github.com>
wrote:

> You need to send notification from the back-end and it will show up
> automatically. Just need to add new type.
>
> As for button, I guess we can add without saving, but then it should be
> OFF by default.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/tananaev/traccar/pull/2104#issuecomment-232567430>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AH9fFyGlM8e3qYY60kmyagZNMe98DpD2ks5qVdYXgaJpZM4JKXy->
> .
>

tananaev commented:
@Abyss777, can you please help with adding new notifications.

Abyss777 commented:
@ninioe you just need to add new event type started from "TYPE_" here https://github.com/tananaev/traccar/blob/master/src/org/traccar/model/Event.java#L50
and it will be in list

You need to implement new event handler. You can use `CommandResultEventHandler` as example https://github.com/tananaev/traccar/blob/master/src/org/traccar/events/CommandResultEventHandler.java
It reacts on one attribute of `position` (just you need)

Then add it to pipeline here https://github.com/tananaev/traccar/blob/master/src/org/traccar/BasePipelineFactory.java#L146
and here
https://github.com/tananaev/traccar/blob/master/src/org/traccar/BasePipelineFactory.java#L206

ninioe commented:
Thanks Abyss777,
I will try this later...

ninioe commented:
@Abyss777, do you think I should create a general AlertEventHandler?

Abyss777 commented:
@ninioe There are some sensors that need to be alerted, may be `SensorsEventHandler` will be better name.

tananaev commented:
I think `AlertEventHandler` sounds good. Some devices report alert type in a message. That's what we want to use. We are not going to generate our own alerts from sensor data. Also, not alerts are sensor alerts. It might be speeding or no GPS or something like that.

condeitokako commented:
this it will rise in the next version ? regards

tananaev commented:
Not sure now. @ninioe, any updates?

ninioe commented:
Yes, I have finished working on it, it works perfect.
I will update the code on GitHub now...
![untitled](https://cloud.githubusercontent.com/assets/8347415/16899208/f2d24e00-4c03-11e6-97af-69f0a7e944c4.png)

tananaev commented:
Please add only English language file. Everything else is automatically generated.

ninioe commented:
OK, i will remember for the next time

On Sun, Jul 17, 2016 at 1:01 PM, Anton Tananaev <notifications@github.com>
wrote:

> Please add only English language file. Everything else is automatically
> generated.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tananaev/traccar/pull/2104#issuecomment-233174370>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AH9fF3gGy-3shrF-B3jgo_5CoQxQbiv6ks5qWf15gaJpZM4JKXy->
> .
>

tananaev commented:
Please revert those language files because it makes diff bigger and harder to read.

ninioe commented:
how do I revert only the language files?

also when I try to revert I get this error:
![image](https://cloud.githubusercontent.com/assets/8347415/16900881/d56a11c2-4c3a-11e6-92a5-edb80c7714f2.png)

ninioe commented:
the revert massed up my code :-(

tananaev commented:
Just overwrite files with the original ones and commit it. You don't need to use git commands to revert it.

tananaev commented:
Also, I looked through the code, and most of my original comments are still not fixed.

tananaev commented:
I would recommend to split pull request into separate ones by feature.

ninioe commented:
Hi Anton,
I overwrite the language files with the original ones.

1. how do I split pull request into separate ones?
2. what comments are still not fixed?

thanks
Erez

tananaev commented:
1. There are basically two options. One is using GitHub cherry-pick commits if you committed code separately for each feature. Another one is create a new branch and just copy files over from your current branch.

2. Please see my comments. There were a lot of the comments that you missed out. For example this line still has a comma:
```java
params = params + (('alarm', 'true'),)
```

ninioe commented:
I can't find this line on my code

tananaev commented on tools/test-generator.py:
```diff
@@ -32,8 +32,10 @@
+def send(lat, lon, course, alarm):
     params = (('id', id), ('timestamp', int(time.time())), ('lat', lat), ('lon', lon), ('bearing', course))
+    if alarm:
+        params = params + (('alarm', 'sosAlarm'),)
```
It is here. Please read original comments. I already did a code review and you basically ignored half of my comments.

ninioe commented:
found it, I don't see any comma

ninioe commented:
I will go over your your comments and check it again...

tananaev commented:
Here is the comma:
```java
params = params + (('alarm', 'sosAlarm'),)
                                        ^
```
Why do we need it there?

tananaev commented:
I would still recommend to split your pull request into smaller independent pieces first.

ninioe commented:
1. I don't see any untreated comment except for the beep function based on base64 string, that I will replace later as soon that I get a beep sound file.
2. To write a tuple containing a single value you have to include a comma, even though there is only one value: http://www.tutorialspoint.com/python/python_tuples.htm
3. why do I still see the language files in the compare?
4. how can I do new pull request to compare only the latest changes?

tananaev commented:
1. OK
2. OK
3. I guess because you haven't committed or pushed your update
4. Not sure what you mean. Pull request should include all the differences between your branch and upstream repo

ninioe commented:
you are right, I forgot to sync after I did the commit
please check it now

tananaev commented:
It looks better, but you still haven't addressed all the comments. Also, can you please all commented out code and unnecessary comments.

ninioe commented:
no comments at all?

tananaev commented:
Only comments that make sense. For example, comment for `setMinMemoryMappedContentLength` makes sense. Another example where comment doesn't make sense:
```java
// Clean positions history
try {
    Context.getDataManager().clearPositionsHistory();
```

ninioe commented:
1. Ok, I will go over all comments again..
2. I saw in your master branch the alarms const inside the file Position.java, but I thought you wanted to define all the alarms inside the Event.java file as part of the notifications?

ninioe commented:
errors I found:

1. when I add or update device with no group I get the following error (and groups table is not empty):

Cannot add or update a child row: a foreign key constraint fails ("traccar"."devices", CONSTRAINT "fk_device_group_groupid" FOREIGN KEY ("groupid") REFERENCES "groups" ("id") ON DELETE SET NULL) - MySQLIntegrityConstraintViolationException (... < QueryBuilder:435 < DataManager:387 < DeviceResource:60 < ...)

2. also constraint problem when trying to change group name:

Cannot add or update a child row: a foreign key constraint fails ("traccar"."groups", CONSTRAINT "fk_group_group_groupid" FOREIGN KEY ("groupid") REFERENCES "groups" ("id") ON DELETE SET NULL) - MySQLIntegrityConstraintViolationException (... < QueryBuilder:435 < DataManager:471 < GroupResource:73 < ...)

![2016-07-17_14-29-18](https://cloud.githubusercontent.com/assets/8347415/16908112/e006dc04-4cd1-11e6-88c3-43dcda007b38.png)

ninioe commented:
My recommendations:
1. allow the user change the device icon
2. show geofence on main map too
3. add checkbox "remember me" on login screen
4. add predefined commands to send the device
5. return the folder view for Groups in the Devices panel (I thing it was better before)
6. you can add more info for each device like: sim number, plate number, remarks etc...

Thanks
Erez

Abyss777 commented:
Hello, @ninioe
Foreign key constraints was fixed https://github.com/tananaev/traccar/commit/2db5a3df91a76776cc214eca1e1fda6aecd0f36d

ninioe commented:
I already deleted those constraints, can you send me the SQL to return it?

Abyss777 commented:
@ninioe 
````SQL
UPDATE devices SET groupid = NULL WHERE groupid NOT IN (SELECT  id FROM  groups);
ALTER TABLE  devices ADD CONSTRAINT  fk_device_group_groupid FOREIGN KEY ( groupid ) REFERENCES groups (id) ON DELETE SET NULL ON UPDATE RESTRICT;

UPDATE groups SET groupid = NULL WHERE groupid NOT IN (SELECT id FROM (SELECT id FROM  groups) AS groups_ids);
ALTER TABLE  groups ADD CONSTRAINT  fk_group_group_groupid FOREIGN KEY ( groupid ) REFERENCES groups (id) ON DELETE SET NULL ON UPDATE RESTRICT;
````

ninioe commented:
thanks

Abyss777 commented:
I have a question, does H02 device send packets with alarm bit once time or continuously?
For example `TYPE_VIBRATION_ALARM`, is it set in one packet in the beginning or in all packets until vibration stopped?

tananaev commented:
@ninioe, for any suggestions and ideas that are not related to this pull request, please create a separate thread.

As for alarm types, I think decoders should use ones I defined in Position model.

ninioe commented:
Hi Anton,
If the alarms will be defined in the Position model how would it be listed in the Notifications List?

tananaev commented:
The way I see it, they either need to map to specific event types, or they can all map to a one generic type "alarm".

ninioe commented:
What do you prefer?

‫ב-18 ביולי 2016, בשעה 13:40, ‏‏Anton Tananaev ‏<notifications@github.com> כתב/ה:‬

> The way I see it, they either need to map to specific event types, or they can all map to a one generic type "alarm".
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

tananaev commented:
For now I would probably prefer a single type because it's easier and we need to finish this pull request.

ninioe commented:
Ok,
I will do all the necessary changes tomorrow morning...

‫ב-18 ביולי 2016, בשעה 14:14, ‏‏Anton Tananaev ‏<notifications@github.com> כתב/ה:‬

> For now I would probably prefer a single type because it's easier and we need to finish this pull request.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

ninioe commented:
Hi Anton, please check my pull request.
I think it's ready now

tananaev commented on src/org/traccar/BasePipelineFactory.java:
```diff
@@ -29,10 +29,7 @@
-import org.traccar.events.GeofenceEventHandler;
-import org.traccar.events.MotionEventHandler;
-import org.traccar.events.OverspeedEventHandler;
+import org.traccar.events.*;
```
No star imports please. Do checkstyle validation of your code please.

tananaev commented on src/org/traccar/Main.java:
```diff
@@ -35,6 +38,18 @@ public static void main(String[] args) throws Exception {
+            public void run() {
+                try {
+                    Context.getDataManager().clearPositionsHistory();
+                } catch (Exception error) {
```
Does this have to be generic exception?

tananaev commented on src/org/traccar/database/DataManager.java:
```diff
@@ -398,6 +398,33 @@ public void updateLatestPosition(Position position) throws SQLException {
     }
 
+    public void clearPositionsHistory() throws SQLException {
+        //SimpleDateFormat s = new SimpleDateFormat("yyyy-MM-dd");
```
Comment still not removed.

tananaev commented on src/org/traccar/database/DataManager.java:
```diff
@@ -398,6 +398,33 @@ public void updateLatestPosition(Position position) throws SQLException {
+            if(lastUpdate != null){
+
+                Date dateBefore = new Date(lastUpdate.getTime() - histDays * 24 * 3600 * 1000 );
+                //String dt = s.format(dateBefore);
```
Comment not removed.

tananaev commented on debug.xml:
```diff
@@ -320,6 +323,10 @@
     </entry>
 
+    <entry key='database.clearPositionsHistory'>
+        DELETE FROM positions WHERE deviceid = :deviceId and servertime &lt; :serverTime;
```
What happens to `positionId` in the device if it points to removed record? It should be handled.

ninioe replied:
it never delete the last position.
it count days back from the last update date and not from current date:

Date dateBefore = new Date(lastUpdate.getTime() - histDays * 24 * 3600 * 1000 );

tananaev replied:
What if the last position was a month ago and you delete everything older than a week?

tananaev commented on src/org/traccar/database/DataManager.java:
```diff
@@ -398,6 +398,33 @@ public void updateLatestPosition(Position position) throws SQLException {
+            Date lastUpdate = device.getLastUpdate();
+            if(lastUpdate != null){
+
+                Date dateBefore = new Date(lastUpdate.getTime() - histDays * 24 * 3600 * 1000 );
```
Extra space before closing bracket.

tananaev commented on web/app/controller/Root.js:
```diff
@@ -89,6 +89,17 @@ Ext.define('Traccar.controller.Root', {
 
+    beep: function() {
+        if(this.snd == null){
+            this.snd = new Audio("data:audio/wav;base64,//uQRAAAAWMSLwUIYAAsYkXgoQwAEaYLWfkWgAI0wWs/ItAAAGDgYtAgAyN+QWaAAihwMWm4G8QQRDiMcCBcH3Cc+CDv/7xA4Tvh9Rz/y8QADBwMWgQAZG/ILNAARQ4GLTcDeIIIhxGOBAuD7hOfBB3/94gcJ3w+o5/5eIAIAAAVwWgQAVQ2ORaIQwEMAJiDg95G4nQL7mQVWI6GwRcfsZAcsKkJvxgxEjzFUgfHoSQ9Qq7KNwqHwuB13MA4a1q/DmBrHgPcmjiGoh//EwC5nGPEmS4RcfkVKOhJf+WOgoxJclFz3kgn//dBA+ya1GhurNn8zb//9NNutNuhz31f/// ...
```
No separate audio file.

tananaev commented on web/app/controller/Root.js:
```diff
@@ -145,6 +156,21 @@ Ext.define('Traccar.controller.Root', {
                         }
                         text = Strings.eventCommandResult + ': ' + text;
+                    } else if (array[i].type === 'alarm' && data.positions) {
+                        text = 'Alarm';
```
No hardcoded text please.

tananaev commented on web/l10n/en.json:
```diff
@@ -118,6 +118,15 @@
+    "alarmVibration": "vibration",
+    "alarmMovement": "movement",
+    "alarmOverspeed": "overspeed",
+    "alarmFallDown": "fallDown",
```
What is this? It's not a proper English phrase. Is it supposed to be id or displayed value?

tananaev commented:
As you can see even automated check failed for your pull request. Also, I've added comments. You still haven't fixed A LOT OF my original comments. Please make sure you fix ALL THE COMMENTS before asking for review.

ninioe commented:
Done.
Please check again...

ninioe commented:
Hi Anton,
I don't know why does the automated check failed?
https://travis-ci.org/tananaev/traccar/builds/145815283

I get this error:
The command "mvn install -DskipTests=true -Dmaven.javadoc.skip=true -B -V" failed and exited with 1 during .

tananaev commented:
```
[INFO] There is 1 error reported by Checkstyle 6.11.2 with checkstyle.xml ruleset.
[ERROR] src/org/traccar/protocol/H02ProtocolDecoder.java:[71,76] (whitespace) WhitespaceAfter: ';' is not followed by whitespace.
```

tananaev commented:
Please run `jshint` and `jscs` checks for JavaScript code as well. Make sure there are no issues there.

ninioe commented:
what is jshint and jscs? and where is it?

tananaev commented:
It's tools that you need to install on your system. They check JavaScript code. Run them from the `web` folder.

ninioe commented:
I found JSHint here and downloaded it:
https://github.com/jshint/jshint/releases/tag/2.9.2

and JSCS here, but it needs npm to be installed and I don't know what is it and how to use it:
http://jscs.info/overview

I will need time to learn how to install and use those tools...

tananaev commented:
You can install both tools using npm (node package manager). Then just run them from `web` directory like this:
```
jshint .
```
and
```
jscs .
```

ninioe commented:
Done

tananaev commented:
Mostly it looks good now, but there is still one unresolved issue:

> What happens to positionId in the device if it points to removed record? It should be handled.

ninioe commented:
this is not a problem,
the device positionId always points to the last received position.
the clear history will never delete the last position, it count days back from the last position date and not from the system date

tananaev commented:
No, it's not. Time in last position and last update time can be completely different.

ninioe commented:
Hi Anton,
The issue is solved, please merge my pull request.

Thanks
Erez

tananaev commented on src/org/traccar/Main.java:
```diff
@@ -17,9 +17,14 @@
 import java.util.Locale;
 
 public final class Main {
+    static final long CLEAN_DELAY = 10 * 1000;
```
Why do we need this delay?

tananaev commented on web/app/view/Devices.js:
```diff
@@ -57,6 +57,23 @@ Ext.define('Traccar.view.Devices', {
+        },{
+            id: 'showAlarmButton',
+            glyph: 'xf0a2@FontAwesome',
+            tooltip: 'Show Alarms',
```
Another hardcoded string.

tananaev commented on web/app/view/Devices.js:
```diff
@@ -57,6 +57,23 @@ Ext.define('Traccar.view.Devices', {
         }, {
             xtype: 'tbfill'
+        },{
+            id: 'showAlarmButton',
```
I don't like the usage of global ids. This introduces tight coupling.

tananaev commented on web/app/view/Devices.js:
```diff
@@ -57,6 +57,23 @@ Ext.define('Traccar.view.Devices', {
+            tooltip: 'Show Alarms',
+            tooltipType: 'title',
+            pressed : true,
+            enableToggle: true,
```
Why is it enabled by default? Any sound should be disabled by default.

tananaev commented on web/app/controller/Root.js:
```diff
@@ -145,6 +156,24 @@ Ext.define('Traccar.controller.Root', {
+                    } else if (array[i].type === 'alarm' && data.positions) {
+                        alarmKey = 'alarm';
+                        text = Strings[alarmKey];
+                        if (typeof text === 'undefined') {
```
Just use `if (text)` instead.

tananaev commented on web/app/controller/Root.js:
```diff
@@ -145,6 +156,24 @@ Ext.define('Traccar.controller.Root', {
+                                if (typeof data.positions[j].attributes.alarm === 'string' && data.positions[j].attributes.alarm.length >= 2) {
+                                    alarmKey = 'alarm' + data.positions[j].attributes.alarm.charAt(0).toUpperCase() + data.positions[j].attributes.alarm.slice(1);
+                                    text = Strings[alarmKey];
+                                    if (typeof text === 'undefined') {
```
Just use if (text) instead.

tananaev commented on web/app/controller/Root.js:
```diff
@@ -160,7 +189,10 @@ Ext.define('Traccar.controller.Root', {
                     }
                     device = Ext.getStore('Devices').getById(array[i].deviceId);
                     if (typeof device !== 'undefined') {
-                        Ext.toast(text, device.getData().name);
```
Use `device.get('name')` instead.

tananaev commented on web/app/controller/Root.js:
```diff
@@ -160,7 +189,10 @@ Ext.define('Traccar.controller.Root', {
-                        Ext.toast(text, device.getData().name);
+                        if (self.showAlarmSelected()) {
+                            self.beep();
+                            Ext.toast(text, device.getData().name);
```
Use `device.get('name')` instead.

tananaev commented on web/app/controller/Root.js:
```diff
@@ -81,14 +81,25 @@ Ext.define('Traccar.controller.Root', {
     },
 
+    beep: function () {
+        if (this.snd === null) {
```
Please use a proper name for variable.

tananaev commented:
Sorry, but I found some more issues.

ninioe commented:
Hi Anton,
I fixed all the issues.
About the toggle button, I meant it to be used to temporary disable all the notifications and not only to mute the beep sound (I fixed it to be more clear). and this is why I think it should be enabled by default.

Thanks
Erez

tananaev commented on src/org/traccar/Main.java:
```diff
@@ -35,6 +40,18 @@ public static void main(String[] args) throws Exception {
+                    Log.warning(error);
+                }
+            }
+        }, CLEAN_DELAY, CLEAN_PERIOD);
```
Delay is still here. Changing the delay value doesn't fix it.

ninioe replied:
this function scheduleAtFixedRate expect 2 parameters.
do you want me to use zero value for the first one, or use another function?

tananaev replied:
Yes, I want you to use 0.

tananaev commented on web/app/controller/Root.js:
```diff
@@ -81,14 +81,25 @@ Ext.define('Traccar.controller.Root', {
             }
         });
-        Ext.get('attribution').remove();
+        if (Ext.get('attribution') !== null) Ext.get('attribution').remove();
```
I just noticed this line. First of all, call it once to avoid multiple searches. Also, you need to add curly braces here.

tananaev commented:
Not everything is fixed again. You haven't removed delay. Also, I notices small issue in js code.

As for been sound, it MUST BE OFF by default. That's the only acceptable option. I don't want web interface to make any sounds unless user explicitly enabled it.

ninioe commented:
I will change the toggle button to mute sound only...

tananaev commented:
I would recommend to add sound to user settings and save it in the database. Then you can keep toggle to temporary disable notifications.

ninioe commented:
I have done what you asked me

tananaev commented:
Looks good. Can you please just rename it from `nuteButton` to `muteButton`.

ninioe commented:
Sorry for this...
It's done

tananaev commented:
Merged, thanks for pull request. In future it would be easier if you submit separate pull request for each feature.

ninioe commented:
Thanks Anton,
About a separate pull request, I already tried this, but as long as I have an opened pull request I cant open another one.

tananaev commented:
You should be able to do it. Just make different branches.

condeitokako commented:
to use this same but with the GPS103- and TK-103 I can help
